cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/data/CassandraDataLayer.java

```diff
@@ -422,7 +424,23 @@ public boolean readIndexOffset()
 
     protected void initInstanceMap()
     {
-        instanceMap = clusterConfig.stream().collect(Collectors.toMap(SidecarInstance::hostname, Function.identity()));
+        if (tokenPartitioner != null)
+        {
+            instanceMap = tokenPartitioner.ring().instances().stream()
+                    .filter(instance -> datacenter == null || datacenter.equals(instance.dataCenter()))
+                    .map(instance -> instance.nodeName()).distinct()
+                    .map(nodeName -> new SidecarInstanceImpl(nodeName, sidecarClientConfig.effectivePort()))
+                    .collect(Collectors.toMap(SidecarInstance::hostname, Function.identity()));
+        }
+        else
+        {
+            instanceMap = Collections.emptyMap();
+        }
```

Comment:
Should it fail early if `tokenPartitioner` is `null`? It is unexpected to have the `null` value. 
Alternatively, the `if-else` can be removed and only keep the `instanceMap` assignment.